### PR TITLE
fix: ignore invalid entitlement value

### PR DESCRIPTION
**Pull request by chrisgacsal**
## Overview

Validate entitlement value before processing balance threshold events.

## Summary by CodeRabbit

* **New Features**
  * Introduced explicit entitlement value validation and a new exported validation error.

* **Bug Fixes**
  * Stricter entitlement validation: negative or inconsistent entitlement values are now detected and reported.
  * Per-threshold availability and activation logic refined: usage now includes overage, percentage thresholds validate against grants, and balance thresholds preserve prior activation behavior with correctness fixes.

* **Tests**
  * Expanded tests for invalid-entitlement scenarios, boundary cases, and explicit error assertions.

### CODE REVIEW

File: openmeter/notification/consumer/entitlementbalancethreshold.go [hunk 1]
```diff
@@ -400,14 +403,14 @@ func getNumericThreshold(threshold notification.BalanceThreshold, value api.Enti
 		totalGrants = lo.FromPtr(value.TotalAvailableGrantAmount)
 	)
 
-	// If no grants are available, we cannot calculate the threshold value.
-	if totalGrants == 0 {
-		return nil, ErrNoBalanceAvailable
+	// TotalAvailableGrants must be equal with the sum of Balance, Usage and Overage.
+	if b := totalGrants - (usage + overage); b-balance > absoluteZero {
```
Comment (gergely-kurucz-konghq) [inline openmeter/notification/consumer/entitlementbalancethreshold.go]: Nit: Could we use more descriptive variable names here?

```suggestion
	if calculatedBalance := totalGrants - (usage + overage); calculatedBalance - balance > absoluteZero {
```

**Review by gergely-kurucz-konghq — DISMISSED** [review]

**Review by tothandras — APPROVED** [review]

**Review by tothandras — CHANGES_REQUESTED** [review]
Reviewing.

File: openmeter/notification/consumer/entitlementbalancethreshold.go [hunk 2]
```diff
@@ -400,14 +403,13 @@ func getNumericThreshold(threshold notification.BalanceThreshold, value api.Enti
...
-		return nil, ErrNoBalanceAvailable
+	if balance < 0 || usage < 0 || overage < 0 || totalGrants < 0 {
+		return nil, fmt.Errorf("%w: negative value is not allowed [balance=%f, usage=%f, overage=%f, total_grants=%f]", ErrInvalidEntitlementValue, balance, usage, overage, totalGrants)
 	}
 
-	// Invalid entitlement value as there cannot be overage if the balance is not zero.
-	if balance > absoluteZero && overage > absoluteZero {
-		return nil, errors.New("balance and overage cannot be positive number at the same time")
+	// TotalAvailableGrants must be equal or greater than the sum of Balance, Usage and Overage.
+	if balance+usage+overage < totalGrants {
```
Comment (tothandras) [inline openmeter/notification/consumer/entitlementbalancethreshold.go]: Comment is incorrect?

**Review by tothandras — DISMISSED** [review]

**Review by tothandras — DISMISSED** [review]

File: openmeter/notification/consumer/entitlementbalancethreshold.go [hunk 3]
```diff
@@ -400,14 +403,13 @@ func getNumericThreshold(threshold notification.BalanceThreshold, value api.Enti
...
-		return nil, ErrNoBalanceAvailable
+	if balance < 0 || usage < 0 || overage < 0 || totalGrants < 0 {
+		return nil, fmt.Errorf("%w: negative value is not allowed [balance=%f, usage=%f, overage=%f, total_grants=%f]", ErrInvalidEntitlementValue, balance, usage, overage, totalGrants)
 	}
 
-	// Invalid entitlement value as there cannot be overage if the balance is not zero.
-	if balance > absoluteZero && overage > absoluteZero {
-		return nil, errors.New("balance and overage cannot be positive number at the same time")
+	// TotalAvailableGrants must be less than or equal to the sum of Balance, Usage and Overage.
+	if balance+usage+overage < totalGrants {
```
Comment (GAlexIHU) [inline openmeter/notification/consumer/entitlementbalancethreshold.go]: why? its not trivial to me that that should be true
